src/Pwa/PageResult/AbstractPageResultHydrator.php

```diff
@@ -0,0 +1,82 @@
+<?php
+
+namespace SwagShopwarePwa\Pwa\PageResult;
+
+use Shopware\Core\Content\Category\CategoryEntity;
+use Shopware\Core\Content\Category\Service\CategoryBreadcrumbBuilder;
+use Shopware\Core\Content\Seo\SeoUrl\SeoUrlEntity;
+use Shopware\Core\Framework\DataAbstractionLayer\EntityRepositoryInterface;
+use Shopware\Core\Framework\DataAbstractionLayer\Search\Criteria;
+use Shopware\Core\Framework\DataAbstractionLayer\Search\Filter\EqualsAnyFilter;
+use Shopware\Core\Framework\DataAbstractionLayer\Search\Filter\EqualsFilter;
+use Shopware\Core\System\SalesChannel\SalesChannelContext;
+use Shopware\Storefront\Framework\Routing\Router;
+use SwagShopwarePwa\Pwa\Controller\PageController;
+
+abstract class AbstractPageResultHydrator
+{
+    /**
+     * @var EntityRepositoryInterface
+     */
+    private $seoUrlRepository;
+
+    /**
+     * @var Router
+     */
+    private $router;
+
+    /**
+     * @var CategoryBreadcrumbBuilder
+     */
+    private $categoryBreadcrumbBuilder;
+
+    public function __construct(Router $router, EntityRepositoryInterface $seoUrlRepository, CategoryBreadcrumbBuilder $categoryBreadcrumbBuilder)
+    {
+        $this->router = $router;
+        $this->seoUrlRepository = $seoUrlRepository;
+        $this->categoryBreadcrumbBuilder = $categoryBreadcrumbBuilder;
+    }
+
+    protected function getBreadcrumbs(CategoryEntity $category, SalesChannelContext $context): array
+    {
+        $breadcrumbs = [];
+
+        $rootCategoryId = $context->getSalesChannel()->getNavigationCategoryId();
+
+        $categoryBreadcrumbs = $this->categoryBreadcrumbBuilder->build($category, null, $rootCategoryId);
+
+        $canonicalUrls = $this->getCanonicalUrls(array_keys($categoryBreadcrumbs), $context);
+
+        foreach ($categoryBreadcrumbs as $id => $name) {
+            $breadcrumbs[$id] = [
+                'name' => $name,
+                'path' => $canonicalUrls[$id] ?? $this->router->generate(PageController::NAVIGATION_PAGE_ROUTE, ['navigationId' => $id]),
```

Comment:
This fallback will not work for product pages, because we're filtering for only navigation page routes.

Reply:
A test making sure, that your breadcrumbs also have proper paths (routes) when no routes have been generated would be required here.

Reply:
Only the breadcrumbs from the seoCategory of a product are returned. The product is not included in the breadcrumbs

```php
# ProductPageResultHydrator.php

if ($seoCategory = $product->getSeoCategory()) {
  $pageResult->setBreadcrumb($this->getBreadcrumbs($seoCategory, $pageLoaderContext->getContext()));
}
```

Reply:
My bad, thanks for noting that. Haven't worked on it for too long (which is ultimately a sign that it's working properly :D)